utils/check_code_style.py

```diff
@@ -0,0 +1,542 @@
+	"^(dynamic|const)_cast<": "C-style typecasts are not to be used",
+	# Matches any '(void)' arguments in methods
+	"\\(void\\)": "do not use void to denote a function with no arguments"
+}
+
+# Patterns for excluding matches (test()#match) of 'include'
+match_exclude = [
+	# Matches any repeating +, - or : operators, or any ::* or ::& references
+	"^.?([+:-])\\1+|::&|::\\*.?$",
+	# Matches any matches which have a -> operator surrounded by at most 1 character on either side.
+	"^.?->.?$",
+	# Matches any matches which have a character followed by '*(' or '&('
+	"^\\w[*&]\\($",
+	# Matches any exponent-related matches.
+	"^e[+-]\\d+$"
+]
+# Patterns for excluding segments that had matches in $include
+segment_exclude = [
+	# Matches anything inside '<>'; this is a bit of a hack for getting rid of type-related issues
+	"<.*>",
+	# Matches any visibility modes; these are followed by ':' marks.
+	"^(public|protected|private|default):$"
+]
+
+# Precompiled version of the regexes
+line_include = {re.compile(regex): description for regex, description in line_include.items()}
+segment_include = {re.compile(regex): description for regex, description in segment_include.items()}
+word_include = {re.compile(regex): description for regex, description in word_include.items()}
+match_exclude = [re.compile(regex) for regex in match_exclude]
+segment_exclude = [re.compile(regex) for regex in segment_exclude]
+
+# Precompiled  helper regexes
+after_comment = re.compile("[^\\s#]")
+whitespace_only = re.compile("^\\s*$")
+whitespaces = re.compile("\\s+")
+
+# The number of errors found in all files
+errors = 0
+# The number of warnings found in all files
+warnings = 0
+# The list of errors and warnings for the current file
+error_list = []
+
+
+# A class used for buffering error messages.
+# text: the text where the error originates from
+# file: the path to the file
+# line: the current line number
+# reason: the reason for the error
+class Error(object):
+
+	def __init__(self, text, file, line, reason):
+		self.text = text
+		self.file = file
+		self.line = line
+		self.reason = reason
+		global errors
+		errors += 1
+		global error_list
+		error_list.append(self)
+
+	# Displays the message for this error
+	def print(self):
+		print("Formatting error in file", self.file, "line", self.line.__str__() + ":", self.text.replace('\n', ""))
+		print("\tReason:", self.reason)
+
+
+# A class used for buffering error messages.
+# text: the text where the warning originates from
+# file: the path to the file
+# line: the current line number
+# reason: the reason for the warning
+class Warning(Error):
+
+	def __init__(self, text, file, line, reason):
+		Error.__init__(self, text, file, line, reason)
+
+		global errors
+		errors -= 1
+		global warnings
+		warnings += 1
+
+	def print(self):
+		print("Warning:", self.reason, "in", self.file, "line", self.line.__str__() + ":", self.text.replace('\n', ""))
+
+
+# Checks the format of all source files.
+def check_code_style():
```

Comment:
This script checks every `.cpp` and `.h` file recursively; it doesn't rely on them being in `/source` or `/test`; as long as the CI's working directory isn't changed, this should detect the source files anywhere. The only possible issues is that the path of `main.cpp` is hard-coded in [one case](https://github.com/endless-sky/endless-sky/pull/7147/files#diff-26baf7b85aa0f100725949f89235646addbf1e841440d8f2bd56e86d89e6d3aeR461), since it has some order-dependent include statements and had to be manually excluded. (Those should not appear anywhere else though, if we adhere to our standards.)